src/channel.ts

```diff
@@ -0,0 +1,268 @@
+	}
+
+	private async mkdir(resource: UriComponents): Promise<void> {
+		return this.provider.mkdir(this.transform(resource));
+	}
+
+	private async readdir(resource: UriComponents): Promise<[string, FileType][]> {
+		return this.provider.readdir(this.transform(resource));
+	}
+
+	private async rename(resource: UriComponents, target: UriComponents, opts: FileOverwriteOptions): Promise<void> {
+		return this.provider.rename(this.transform(resource), URI.from(target), opts);
+	}
+
+	private copy(resource: UriComponents, target: UriComponents, opts: FileOverwriteOptions): Promise<void> {
+		return this.provider.copy(this.transform(resource), URI.from(target), opts);
+	}
+
+	private async watch(session: string, req: number, resource: UriComponents, opts: IWatchOptions): Promise<void> {
+		this.watchers.get(session)!._watch(req, this.transform(resource), opts);
+	}
+
+	private async unwatch(session: string, req: number): Promise<void> {
+		this.watchers.get(session)!.unwatch(req);
+	}
+
+	private transform(resource: UriComponents): URI {
+		// HACK: for now assume /out is relative to the build (used for the
+		// walkthrough content).
+		if (resource.path.indexOf("/out") === 0) {
+			resource.path = this.environmentService.appRoot + resource.path;
+		}
+		return URI.from(resource);
+	}
+}
+
+export class ExtensionEnvironmentChannel implements IServerChannel {
+	public constructor(
+		private readonly environment: IEnvironmentService,
+		private readonly log: ILogService,
+		private readonly telemetry: ITelemetryService,
+	) {}
+
+	public listen(_: unknown, event: string): Event<any> {
+		throw new Error(`Invalid listen "${event}"`);
+	}
+
+	public async call(context: any, command: string, args?: any): Promise<any> {
+		switch (command) {
+			case "getEnvironmentData":
+				return transformOutgoingURIs(
+					await this.getEnvironmentData(args.language),
+					getUriTransformer(context.remoteAuthority),
+				);
+			case "getDiagnosticInfo": return this.getDiagnosticInfo();
+			case "disableTelemetry": return this.disableTelemetry();
+		}
+		throw new Error(`Invalid call "${command}"`);
+	}
+
+	private async getEnvironmentData(locale: string): Promise<IRemoteAgentEnvironment> {
+		return {
+			pid: process.pid,
+			appRoot: URI.file(this.environment.appRoot),
+			appSettingsHome: this.environment.appSettingsHome,
+			settingsPath: this.environment.machineSettingsHome,
+			logsPath: URI.file(this.environment.logsPath),
+			extensionsPath: URI.file(this.environment.extensionsPath),
+			extensionHostLogsPath: URI.file(path.join(this.environment.logsPath, "extension-host")),
+			globalStorageHome: URI.file(this.environment.globalStorageHome),
+			userHome: URI.file(this.environment.userHome),
+			extensions: await this.scanExtensions(locale),
+			os: OS,
+		};
+	}
+
+	private async scanExtensions(locale: string): Promise<IExtensionDescription[]> {
+		const translations = {}; // TODO: translations
```

Comment:
Will not having this break language packs?

Reply:
Yeah, the language packs don't work at all yet I believe. I haven't messed with any of that yet. Do you think we should try getting that in for this PR?